lib/pipeline/webpackProcess.js

```diff
@@ -1,208 +1,104 @@
+        fs.dappPath('.embark/embark-aliases.json'),
+        JSON.stringify(importsList)
+      );
+      await writeFile(
+        fs.dappPath('.embark/embark-assets.json'),
+        JSON.stringify(assets)
+      );
+    } catch (e) {
+      return callback(errorMessage(e));
+    }
 
-      function changeCwdBack(next) {
-        process.chdir(realCwd);
-        next();
-      }
-    ], (err) => {
-      process.chdir(realCwd);
-      callback(err);
-    });
-  }
+    const dappConfigPath = fs.dappPath('webpack.config.js');
+    const defaultConfigPath = fs.embarkPath('lib/pipeline', 'webpack.config.js');
 
-  webpackRun(filename, options, includeModules, importsList, detectErrors, realCwd, callback) {
-    const self = this;
-    glob(fs.dappPath('.embark/versions/*/*'), (err, files) => {
-      let versions;
-      if (err) {
-        console.error(err);
-        versions = [];
+    let config, configPath;
+    try {
+      if (fs.existsSync(dappConfigPath)) {
+        configPath = dappConfigPath;
+        delete require.cache[configPath];
       } else {
-        versions = files;
+        configPath = defaultConfigPath;
       }
-      let defaultOptions = {
-        mode: self.env === 'production' ? 'production' : 'none',
-        // devtool: self.env === 'development' ? 'source-map' : false,
-        // pipeline would need to copy .map files to dist/ target dir
-        // note: generating full source maps ('source-map') roughly doubles build time
-        entry: fs.dappPath(filename),
-        output: {
-          globalObject: 'typeof self !== \'undefined\' ? self : this',
-          libraryExport: 'default',
-          libraryTarget: 'umd',
-          path: fs.dappPath('.embark'),
-          filename: filename,
-          umdNamedDefine: true
-        },
-        // profile: true,
-        // stats: 'verbose',
-        // note: generating and writing to disk verbose stats increases build time
-        resolve: {
-          alias: importsList,
-          modules: [
-            fs.dappPath('node_modules'),
-            ...versions,
-            fs.embarkPath('node_modules')
-          ]
-        },
-        plugins: [
-          new HardSourceWebpackPlugin({
-            cacheDirectory: fs.dappPath('node_modules/.cache/hard-source'),
-            // ufglify (wp mode: production) will still save its cache in embark's node_modules/.cache/
-            environmentHash: {
-              root: fs.dappPath()
-            }
-          }),
-          new HardSourceWebpackPlugin.ExcludeModulePlugin(
-            [{test: /app[\\/]|contracts[\\/]/}]
-          )
-        ]
-      };
-
-      let webpackOptions = utils.recursiveMerge(defaultOptions, options);
+      config = require(configPath);
+      // valid config types: https://webpack.js.org/configuration/configuration-types/
+      // + function that returns a config object
+      // + function that returns a promise for a config object
+      // + array of named config objects
```

Comment:
I guess it depends on what you mean by multiple configs.

If you mean multiple formats:  my thinking was that our config-handler should accept any valid webpack config type, in case a dapp author wants to use a different one than I chose for our default `webpack.config.js`.  In other words, if the dapp author ejects our config and rewrites it to export a function, we should be able to load that config.

If you mean multiple configs being exported from our default or a customized config:  that's how the `--pipeline` option provides some flexibility.  If `webpack.config.js` exports and array, it is assumed (in `webpackProcess.js`, per webpack's docs) that each member of the array will be an object that has a `name:` property.  That property is matched up to `--pipeline [name]`. So a dapp author could clone our `base` config (like we do to create the `development` and `production` configs), change some things, give it `name: 'foo'`, add it to the `module.exports` array, then make use of it via `embark build|console|run --pipeline foo`.

Reply:
Yes, fantastic, I can see this being very flexible and will be helpful in the future. It's definitely above and beyond. I didn't get a chance to test it in the ways you've described however.